[ADD] estate: list, buy and sell properties

ijja-odoo opened the pull request:
This PR initializes the Estate application, allowing users to list, buy, and
sell properties. It includes model creation with essential fields, access rules
to manage permissions and basic views for property management.

Commit:
[ADD] estate: add basic view, fields, action and menu

It includes model creation with essential fields, access rules to manage
permissions and basic views for property management.

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/tutorials/pull/252.png)](https://mergebot.odoo.com/odoo/tutorials/pull/252)

aktr-odoo commented on estate/__init__.py:
```diff
@@ -0,0 +1 @@
+from . import models
```
Please leave a new line at the end of the every file

aktr-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,20 @@
+from odoo import models, fields
+from dateutil.relativedelta import relativedelta
```
Import files according to coding guidelines

aktr-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,20 @@
+from odoo import models, fields
+from dateutil.relativedelta import relativedelta
+
+class EstatePropertyModel(models.Model):
```
Give a proper class name.
HINT: Refer codebase

aktr-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,20 @@
+
+class EstatePropertyModel(models.Model):
+    _name = "estate.property"
+    _description = "Estate property model displaying single estate property"
```
Too long description.

aktr-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,20 @@
+    garage = fields.Boolean("Garage")
+    garden = fields.Boolean("Garden")
+    garden_area = fields.Integer("Garden Area")
+    garden_orientation = fields.Selection([("north", "North"), ("south", "South"), ("east", "East"), ("west", "West")],string="Garden Orientation")
```
Format this.

aktr-odoo reviewed (changes requested):
:wave: 

I have left a few comments. Can you please have a look?

Commit:
[IMP] estate: add estate property type

Added estate property type implementing  search with domain, filters and
group by and used many2one field to connect one estate property type with
multiple estate property

Commit:
[IMP] estate: add estate property tag and offer

Completed chapter 7,8 and 9 in which added many2many relations of property and
tags and using one2many relations of multiple offers connecting to one property.
Used computed fields, onchanges for various estate properties and how on button
click python method will be executed based on actions of type object.

Commit:
[IMP] estate: add constraints, widgets and conditional attributes

Added Python and SQL constraints for data validation. Implemented status bar
widget, list ordering and sequence based ordering. Added conditional visibility
for buttons and fields and applied tag colors.

Commit:
[IMP] estate: add inehritance and invoice generation

Added stat button which links multiple offers to single property type. Inherited
res.users model and view to which displays properties linked with corresponding
salesperson. Implemented invoice generation once the property gets to sold state.

Commit:
[IMP] estate: add kanban view

Added a kanban view using QWeb templates and implemented conditional fields for
dynamic field visibility.

aktr-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,135 @@
+        copy=False,
+        default=lambda self: fields.Date.today() + relativedelta(days=90),
+    )
+    expected_price = fields.Float(
+        "Expected Price",
+        required=True,
+        help="Expected Price"
+    )
+    selling_price = fields.Float(
+        "Selling Price",
+        readonly=True,
+        copy=False
+    )
+    bedrooms = fields.Integer(
+        "Bedrooms",
+        default=2
+    )
```
Format this

aktr-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,135 @@
+    
+    def action_mark_property_sold(self):
+        for record in self:
+            if record.state == "cancelled":
```
Do we need this condition?

aktr-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,135 @@
+    
+    def action_mark_property_cancelled(self):
+        for record in self:
+            if record.state == "sold":
```
Do we need this condition?

aktr-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,135 @@
+        return True
+
+    @api.constrains("selling_price", "expected_price")
+    def _check_valid_selling_price(self):
```
Constraints must be defined before compute

aktr-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,74 @@
+    @api.model_create_multi
+    def create(self, vals_list):
+        for vals in vals_list:
+            property = self.env['estate.property'].browse(vals['property_id'])
```
Do we need `self.env` if we directly have `property_id`

aktr-odoo commented on estate/models/inherited_res_users.py:
```diff
@@ -0,0 +1,12 @@
+from odoo import models, fields
+
+
+class InheritedResUsers(models.Model):
```
Give proper class name

aktr-odoo commented on estate_account/models/estate_property.py:
```diff
@@ -0,0 +1,25 @@
+from odoo import models, Command
+
+
+class InheritedEstateProperty(models.Model):
```
Give proper class name

aktr-odoo commented on estate_account/models/estate_property.py:
```diff
@@ -0,0 +1,25 @@
+    _inherit="estate.property"
+    
+    def action_mark_property_sold(self):
+        self.env["account.move"].create({
```
Don't we need to check access rights here?

aktr-odoo reviewed (changes requested):
:wave: 

I have left a few comments. Could you please have a look?

Commit:
[IMP] estate: add groups and access rights

Added manager group having read, write and create access to property model and
full access to rest of the models with that another group agent having read only
access to property type and tag models.

Commit:
[IMP] estate: add rule based accessing

Added various record rules for groups which introduces access only necessary
part of modules, made the settings menu action only visible to managers and not
to agents and refactored code according to coding guidelines.

Commit:
[IMP] estate: add default data and demo data

Added default data for estate property type under data folder and added demo
data for estate property and demo offers linking to that demo properties.

Commit:
[IMP] estate: add print reports

Added QWeb reports and templates to dynamically generate and print PDFs,
using basic reports, sub-templates, and inheritance.

aktr-odoo commented on estate/security/security.xml:
```diff
@@ -0,0 +1,44 @@
+    </record>
+    
+    <record id="property_visibility_manager_rule" model="ir.rule">
+        <field name="name">An agent manager manages property types, tags, and oversees all properties.</field>
```
Rule's names are quite big.
Refer codebase to give proper naming.

Commit:
[IMP] estate: add tests

I have added a tests folder with estate property tests. These tests cover the
basic functionality of properties with that added chatter in estate property
form view and helper for empty views.

Commit:
[IMP] estate: add wizard, image field, and controller with pagination

Added a wizard to make a common offer for multiple selected properties. Added an
image field to the estate property model. Implemented a controller to list all
available properties on the website with a property detail view and  pagination.

Commit:
[IMP] awesome_owl: add components, props and states

Added Counter and Card components with prop handling and function binding,
including optional props with validation, parent to child data passing via
props, child to parent communication using callbacks, and function binding to
maintain the correct this context.

Commit:
[IMP] awesome_*: add todo list, generic card, and statistics service

Completed awesome_owl chapter by adding a basic todo list with add, delete and
mark complete functionality using dynamic attributes and input events with that
modified the card component to be generic by using slots in owl and adding
toggle button to show and hide card body. Started awesome_dashboard chapter in
that added the already available layout component learned about services and how
to register and call them, how to open existing views using action service, made
generic dashboard item with variable size prop, route call with rpc and created
and registered statistics service which loads orders statistics.

Commit:
[IMP] awesome_dashboard: add pie chart, real time updates and lazy loading

Added a pie chart to the dashboard and modified the statistics service to run
at fixed intervals. Used a reactive object for dynamic updates and converted the
dashboard into a lazy component.

Commit:
[IMP] awesome_dashboard: add dashboard config dialog and made dashboard generic

Made dashboard items generic now able to take components, added dashboard items
to awesome_dashboard registry and later importing all items of registry in the
dashboard component instead of exporting all the items. Added dashboard items
configuration dialog through which dashboard items can be dynamically set
visible or not storing it in browser's local storage. Added module icon and made
the dashboard items card taking full width in small screens.

aktr-odoo reviewed (changes requested):
:wave: 

I have left a comment. Could you please have a look?

ijja-odoo commented on awesome_owl/static/src/card/card.js:
```diff
@@ -0,0 +1,17 @@
+    static template = "awesome_owl.Card"
+    static props = {
+        title: { type: String, optional: false},
+        slots: { type: Object, optional: true }
```
Yes sir, if I remove slots from the props an error will be shown in the browser's console.

![Error Ref](https://github.com/user-attachments/assets/34cf3d3f-707d-48ca-8bf8-ebfe64caf499)

aktr-odoo replied:
Can you please explain the reason of using slots here?

aktr-odoo commented on event_ticket_limiter/views/event_templates_page_registration.xml:
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" ?>
+<odoo>
+    <template id="modal_ticket_registration" inherit_id="website_event.modal_ticket_registration">
+        <xpath expr="//t[@t-set='seats_max_ticket']" position="replace">
```
Replacing `t-set` is not a good option. Try it in another way.
Follow the same for all xpath.

aktr-odoo reviewed (changes requested):
:wave: 

I have left a few comments for `event_ticket_limiter`. Could you please have a look?

ijja-odoo commented on event_ticket_limiter/models/event_ticket.py:
```diff
@@ -0,0 +1,21 @@
+    @api.onchange("max_tickets_per_registration")
+    def _onchange_max_tickets_per_registration(self):
+        """Ensures max_tickets_per_registration is at least 1 and does not exceed available seats if limited."""
+        if self.max_tickets_per_registration <= 0:
+            raise UserError("Atleast 1 ticket must be set per registration")
```
You're right sir! The check should be in @api.constrains instead of @api.onchange, corrected it.

aktr-odoo commented on event_ticket_limiter/views/event_templates_page_registration.xml:
Give proper file name

ijja-odoo commented:
@aktr-odoo Requested changes have been made can you please have a look?

aktr-odoo commented on event_ticket_limiter/models/event_ticket.py:
```diff
@@ -0,0 +1,22 @@
+    def _check_max_tickets_per_registration(self):
+        """Ensures max_tickets_per_registration is at least 1 and does not exceed available seats if limited."""
+        for record in self:
+            if record.max_tickets_per_registration <= 0:
+                raise ValidationError("At least 1 ticket must be set per registration.")
```
Any reason to do this using python constraints?

ijja-odoo replied:
If I add this as a SQL constraint there are some existing values which violates the condition, which in turn returns the below warning in the terminal:

![image](https://github.com/user-attachments/assets/5536a6de-c0ab-4507-b8e4-5640af95995d)

Reason:
![image](https://github.com/user-attachments/assets/7f5f91f7-67ea-4274-9e8a-1cef4ecb5c11)
In the database table there are some values violating the condition.

Due to that, I add the check as python constrains,

aktr-odoo replied:
> If I add this as a SQL constraint there are some existing values

But why are there existing (pre-defined) values in DB? I can't see any demo data that has defined `max_tickets_per_regestration`

ijja-odoo replied:
Okay, so the reason is before adding the constraints I have added the field and changed data for testing purpose, now after creating the new database I am not getting any kind of warning related to the constraints. So I will move the check from python constrains to sql constraints.

aktr-odoo reviewed (changes requested):
:wave: 

I have left a few comments. Could you please have a look?

ijja-odoo commented:
@aktr-odoo Requested changes have been made can you please have a look?

Commit:
[ADD] estate_auction: add automated auction system for estate

PURPOSE

The purpose of this commit is to introduce an automated auction system for
estate properties. This feature enhances the property sales process by enabling
auction based transactions, allowing users to bid on properties and automate
property sales process.

SPECIFICATION

- This introduces an automated auction system for estate properties. Users can
  now list properties for auction, allowing buyers to place bids now from
  website.
- Once a property is sold through auction, an automated email notification is
  sent to the users. The system also includes an automatic process that runs
  every five minutes to check if an auction has ended and updates the property
  status accordingly.
- Additionally, new options have been added to manage the property sale mode,
  including buttons to start an auction, track invoices related to the
  properties, and configure auction settings directly from the property form.

Task-4625884
